xdcr.go

```diff
@@ -0,0 +1,245 @@
+// the file licenses/APL2.txt.
+
+package rosmar
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"sync/atomic"
+
+	sgbucket "github.com/couchbase/sg-bucket"
+)
+
+// XDCR implements a XDCR bucket to bucket setup within rosmar.
+type XDCR struct {
+	terminator              chan bool
+	fromBucketCollectionIDs map[uint32]sgbucket.DataStoreName
+	fromBucket              *Bucket
+	toBucket                *Bucket
+	replicationID           string
+	docsWritten             atomic.Uint64
+	docsFiltered            atomic.Uint64
+	errorCount              atomic.Uint64
+}
+
+// NewXDCR creates an instance of XDCR backed by rosmar. This is not started until Start is called.
+func NewXDCR(_ context.Context, fromBucket, toBucket *Bucket, opts sgbucket.XDCROptions) (*XDCR, error) {
+	if opts.Mobile != sgbucket.XDCRMobileOn {
+		return nil, errors.New("Only sgbucket.XDCRMobileOn is supported in rosmar")
+	}
+	return &XDCR{
+		fromBucket:              fromBucket,
+		toBucket:                toBucket,
+		replicationID:           fmt.Sprintf("%s-%s", fromBucket.GetName(), toBucket.GetName()),
+		fromBucketCollectionIDs: map[uint32]sgbucket.DataStoreName{},
+		terminator:              make(chan bool),
+	}, nil
+
+}
+
+// getFromBucketCollectionName returns the collection name for a given collection ID in the from bucket.
+func (r *XDCR) getFromBucketCollectionName(collectionID uint32) (sgbucket.DataStoreName, error) {
+	dsName, ok := r.fromBucketCollectionIDs[collectionID]
+	if ok {
+		return dsName, nil
+	}
+
+	dataStores, err := r.fromBucket.ListDataStores()
+
+	if err != nil {
+		return nil, fmt.Errorf("Could not list data stores: %w", err)
+
+	}
+
+	for _, dsName := range dataStores {
+		dataStore, err := r.fromBucket.NamedDataStore(dsName)
+		if err != nil {
+			return nil, fmt.Errorf("Could not get data store %s: %w", dsName, err)
+		}
+
+		if dataStore.GetCollectionID() == collectionID {
+			name := dataStore.DataStoreName()
+			r.fromBucketCollectionIDs[collectionID] = name
+			return name, nil
+
+		}
+	}
+	return nil, fmt.Errorf("Could not find collection with ID %d", collectionID)
+}
+
+// Start starts the replication.
+func (r *XDCR) Start(ctx context.Context) error {
+	args := sgbucket.FeedArguments{
+		ID:         "xdcr-" + r.replicationID,
+		Backfill:   sgbucket.FeedNoBackfill,
+		Terminator: r.terminator,
+	}
+
+	callback := func(event sgbucket.FeedEvent) bool {
+		docID := string(event.Key)
+		trace("Got event %s, opcode: %s", docID, event.Opcode)
+		dsName, err := r.getFromBucketCollectionName(event.CollectionID)
+		if err != nil {
+			warn("Could not find collection with ID %d for docID %s: %s", event.CollectionID, docID, err)
+			r.errorCount.Add(1)
+			return false
+
+		}
+
+		switch event.Opcode {
+		case sgbucket.FeedOpDeletion, sgbucket.FeedOpMutation:
+			if strings.HasPrefix(docID, sgbucket.SyncDocPrefix) && !strings.HasPrefix(docID, sgbucket.Att2Prefix) {
```

Comment:
It would be nice to abstract this as a key filter function that's defined on the replication (i.e. XDCR.keyFilterFunc), and set that filter to a pre-defined function when the replication is created with sgbucket.XDCRMobileOn.  That would keep the general XDCR implementation a bit more generic and make it clear what's happening when XDCRMobileOn is set.